fairworkflows/fairworkflow.py

```diff
@@ -33,6 +34,39 @@ def __init__(self, description, uri=None):
         self._steps = {}
         self._last_step_added = None
 
+    @classmethod
+    def from_rdf(cls, rdf: rdflib.Graph, uri: str = None):
+        """Construct Fair Workflow from existing RDF.
+
+        Args:
+            rdf: RDF graph containing information about the workflow AND it's
+                associated steps. Should use pplan ontology.
+            uri: URI of the workflow
+        """
+        self = cls(uri)
+        self._rdf = copy(rdf)  # Make sure we don't mutate user RDF
+        if rdflib.URIRef(self._uri) not in rdf.subjects():
+            warnings.warn(f"Warning: Provided URI '{self._uri}' does not "
+                          f"match any subject in provided rdf graph.")
+        self._extract_steps_from_rdf()
+        self.anonymise_rdf()
+        return self
+
+    def _extract_steps_from_rdf(self):
+        """Extract FairStep objects from self.rdf.
+
+        Extract steps from RDF. Removes triples describing steps from self.rdf.
+        """
+        step_refs = self.rdf.subjects(predicate=Nanopub.PPLAN.isStepOfPlan,
+                                      object=self.self_ref)
+        for step_ref in step_refs:
+            step_rdf = rdflib.Graph()
+            for triple in step_rdf.triples((step_ref, None, None)):
+                step_rdf.add(triple)
+                self.rdf.remove(triple)
+            step = FairStep.from_rdf(self.rdf, uri=str(step_ref))
+            self._steps[step.uri] = step
+
```

Comment:
So this assumes that you pass it RDF with both the workflow and the steps in it. I am completely in the blue whether that is the way a user would want to interact with this. But it seemed the most logical way, because there is not always references to steps in the workflow RDF, whereas there is in the step RDF.

Reply:
When you say it assumes the workflow RDF "contains" the steps you mean all the triples describing the step are provided in this block of RDF, in which case it converts the triples into a ```FairStep```? I think that is indeed something that can happen (i.e. a likely use of the library). As usual in the semantic world it is also likely that the workflow will describe itself entirely in terms of steps using URIs that are not contained in this block of RDF. Ideally we would handle the case where there is a mixture of steps, some defined in the provided RDF, and some just URI references that may be loaded later if required.

I guess one problem with not loading the step rdf for 'externally defined' steps would be that validation of inputs/outputs might not then be possible. Would it make sense to instead implement a sort of 'lazy loading' of FairSteps? We could have a local cache of ```FairStep``` objects, essentially like the ```_steps``` dict that already exists, except that if a FairStep is requested that is not in the cache, the lib attempts to fetch that step's RDF from whatever its given URI is.

That's one way of doing this, I suppose. The alternative would be automatically fetching any RDF from any URI that is not already contained within the workflow's RDF. I think my fear with that is that we somehow end up in an eternal loading walk down the semantic graph (especially as a step could be a ```Plan``` itself, in general).

Reply:
Ok, so there's two situations then:

1. All steps are defined in the provided RDF. This is solved with this PR
2. Some or all steps in the workflow RDF actually point to steps not defined in the provided RDF.

I will build something that also solves case 2. This will look like you described it. I think it is indeed wise to do some lazy loading. I guess that we for now will only need to load the step itself. Even if the step is a Plan I don't think we need fetch everything that it is pointing to. And let's not worry about that in this phase! :)

Reply:
There's a problem. The workflow does not point to the steps that are part of it, the steps point to the workflow by the [p-plan:isStepOfPlan](https://www.opmw.org/model/p-plan/#isStepOfPlan) predicate. Ways to solve it:
* Pass a list of uris of steps which we will fetch
* Assume that the dul:precedes predicate and http://purl.org/spar/pwo/hasFirstStep together always describe all the steps.
* Assume that workflows point to steps with: http://purl.org/spar/pwo/hasStep 

BTW why do we use dul:precedes and not https://www.opmw.org/model/p-plan/#isPreceededBy?

Reply:
I misunderstood that https://www.opmw.org/model/p-plan/#isStepOfPlan is actually part of the workflow description, and not the step.

Reply:
Implemented this

Reply:
Good work on this @svenvanderburg . This was far trickier than it first appeared.